deployments/helm/nvidia-device-plugin/templates/daemonset-gfd.yml

```diff
@@ -57,7 +57,31 @@ spec:
       {{- end }}
       {{- if $options.hasConfigMap }}
       shareProcessNamespace: true
+      {{- end }}
       initContainers:
+      - image: ubuntu
+        name: gpu-feature-discovery-imex-init
+        command:
+        - "/bin/bash"
+        - "-c"
+        - |
+          if [[ ! -d /driver-root/etc/nvidia-imex ]]; then
+            echo "No IMEX Nodes config path detected; Skipping"
+            exit 0
+          fi
+
+          mkdir -p /config/etc/nvidia-imex
+          mount --bind /driver-root/etc/nvidia-imex /config/etc/nvidia-imex
+        volumeMounts:
+          - name: config
+            mountPath: /config
+            mountPropagation: Bidirectional
```

Comment:
Why does this need to be bidirectional? At worst I think we only need `HostToContainer `-- but I'm curious if we even care about future mounts in either direction?

Reply:
The issue I was seeing was that the mount made in the init container was not visible to other containers using the empty volume. Maybe there is a better way to ensure that the mounts are shared than propagating this back to the empty volume on the host.

Reply:
OK, yeah, that makes sense. The init container creates the mount, but if this mount is not reflected back on the host, then when it exits, the mount will be removed and never visible to the app container when it eventually starts up. Let's leave it this way for now and think a bit more about this for future releases.

Reply:
The one issue that I did see was that the mount was not cleaned up. What is the best practice for doing this? (we don't do this for MPS either ... )

Reply:
Where does it live after the pod has been shut down? I would assume the shutdown of the pod and the removal of the temporary space for the mounted volume would have cleaned it up.

Reply:
Running `mount | grep config` in the kind worker node container:
```
 mount  | grep config
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
overlay on /var/lib/kubelet/pods/406054da-7ef0-48f1-a4ad-1da282913dc7/volumes/kubernetes.io~empty-dir/config/etc/nvidia-imex type overlay (ro,relatime,lowerdir=/var/lib/docker/overlay2/l/YJ3LUQCUVC2MXFPQQU5LOFINP2:/var/lib/docker/overlay2/l/NJLKU2BR3I52CNQBK6SCGI74WJ:/var/lib/docker/overlay2/l/BXAROID4RF5JY7PFABBT42PHK4,upperdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/diff,workdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/work,xino=off)
```

immediately after deleting the pod we see:
```
nvidia-device-plugin   nvidia-device-plugin-gpu-feature-discovery-wfh2f                  0/1     Terminating   0               29m
```
and
```
mount  | grep config
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
overlay on /var/lib/kubelet/pods/406054da-7ef0-48f1-a4ad-1da282913dc7/volumes/kubernetes.io~empty-dir/config/etc/nvidia-imex type overlay (ro,relatime,lowerdir=/var/lib/docker/overlay2/l/YJ3LUQCUVC2MXFPQQU5LOFINP2:/var/lib/docker/overlay2/l/NJLKU2BR3I52CNQBK6SCGI74WJ:/var/lib/docker/overlay2/l/BXAROID4RF5JY7PFABBT42PHK4,upperdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/diff,workdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/work,xino=off)
overlay on /var/lib/kubelet/pods/3a948d67-b23b-41d3-a172-18ec8b37e7ad/volumes/kubernetes.io~empty-dir/config/etc/nvidia-imex type overlay (ro,relatime,lowerdir=/var/lib/docker/overlay2/l/YJ3LUQCUVC2MXFPQQU5LOFINP2:/var/lib/docker/overlay2/l/NJLKU2BR3I52CNQBK6SCGI74WJ:/var/lib/docker/overlay2/l/BXAROID4RF5JY7PFABBT42PHK4,upperdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/diff,workdir=/var/lib/docker/overlay2/f94a9595453a2c4a09dfaf3720ef1beb376620eec59443e7cbb4743e1a9d59ef/work,xino=off)
```
running `umount` on the worker node allows terminations to continue / succeed.

Reply:
Should we just go back to a copying approach instead of a mounting? It means we will require a restart of the plugin if the config file changes on the host, but this is probably OK for this first iteration.

Reply:
It's longer than 30s. I have one that's in terminating for more than 6 hours.

It could be that using the mounting logic that we have for the mps daemon doesn't have this issue since there is more complex logic being applied?

Reply:
For what it's worth, I'm happy to resort to copying and require a restart, alternatively we can do something like:
```
diff --git a/deployments/helm/nvidia-device-plugin/templates/daemonset-gfd.yml b/deployments/helm/nvidia-device-plugin/templates/daemonset-gfd.yml
index cb3f9d36..fd1ffb8a 100644
--- a/deployments/helm/nvidia-device-plugin/templates/daemonset-gfd.yml
+++ b/deployments/helm/nvidia-device-plugin/templates/daemonset-gfd.yml
@@ -214,6 +214,9 @@ spec:
         {{- end }}
           - name: config
             mountPath: /config
+          - name: driver-root-etc
+            mountPath: /driver-root/etc
+            readOnly: true
         {{- with .Values.resources }}
         resources:
           {{- toYaml . | nindent 10 }}
@@ -225,9 +228,9 @@ spec:
         - name: host-sys
           hostPath:
             path: "/sys"
-        - name: driver-root
+        - name: driver-root-etc
           hostPath:
-            path: {{ clean ( join "/" ( list "/" .Values.nvidiaDriverRoot ) ) | quote }}
+            path: {{ clean ( join "/" ( list "/" .Values.nvidiaDriverRoot "etc" ) ) | quote }}
       {{- if $options.hasConfigMap }}
         - name: available-configs
           configMap:
```
but this doesn't address the operator case?

Reply:
It could work in the operator case if we updated both the host-root and drive-root mounts to `/etc` --> `/host/etc` and `/run/nvidia/driver/etc` --> `/driver-root/etc` respectively.